### PR TITLE
Avoid reloading font collection for spawned engines with compatible asset managers

### PR DESCRIPTION
**Pull request by dnfield**
@jason-simmons @jiahaog fyi

I need to figure out a test for this still but it seems to work. Hot restart is kind of a mess though, because `Engine::Restart` nulls out the asset manager it ends up reloading the font collection for every engine after a restart. But we separately need to fix hot restart, it's not very usable on the example app 

Fixes https://github.com/flutter/flutter/issues/143701

### CODE REVIEW

**Comment by jiahaog** [comment]
Nice, just tested this (two engines) and it seems to solve the issue:

```
pprof -top  /tmp/heap_profile-latest/*.gz | grep createTypeface
```

```
# Before
0     0%   100%     3.15MB 11.29%  flutter::AssetManagerFontStyleSet::createTypeface(int)

# After
0     0%   100%     1.57MB  6.04%  flutter::AssetManagerFontStyleSet::createTypeface(int)
```

**Review by jiahaog — APPROVED** [review]
Looks great, but I'm not so familiar with the code and perhaps Jason should review this too.

File: assets/asset_resolver.h [hunk 1]
```diff
@@ -81,6 +93,10 @@ class AssetResolver {
     return {};
   };
 
+  virtual bool operator==(const AssetResolver& other) const = 0;
+
+  bool operator!=(const AssetResolver& other) { return !operator==(other); }
```
Comment (jason-simmons) [inline assets/asset_resolver.h:98]: nit: declare this method as `const`

File: shell/common/engine.cc [hunk 2]
```diff
@@ -174,7 +175,8 @@ fml::WeakPtr<ImageGeneratorRegistry> Engine::GetImageGeneratorRegistry() {
 
 bool Engine::UpdateAssetManager(
     const std::shared_ptr<AssetManager>& new_asset_manager) {
-  if (asset_manager_ == new_asset_manager) {
+  if ((asset_manager_ && new_asset_manager) &&
```
Comment (jason-simmons) [inline shell/common/engine.cc:178]: nit: the parentheses around `(asset_manager_ && new_asset_manager)` are not needed

**Review by jason-simmons — APPROVED** [review]

File: shell/common/engine.cc (hunk 2 above)
Comment (dnfield) [reply shell/common/engine.cc:178]: Done

File: assets/asset_resolver.h (hunk 1 above)
Comment (dnfield) [reply assets/asset_resolver.h:98]: Done

**Pull request merged**